### DISCUSSION TITLE
Is anyone building a file explorer with textual?

### DISCUSSION THREAD

**Discussion by marcbelmont** [Textualize/textual#2199] [General]
I'm a long time user of [Ranger](https://github.com/ranger/ranger/) (a file explorer written in Python), but development has stalled (last release is 3 years old) and a reboot using textual seems like a fun project. 

I know very little about textual. I would love to get advice for how to architecture such a project and if it makes sense to use this library.
Is anyone working on a file explorer already?

**Comment by willmcgugan** [comment]
Our very own @darrenburns was building a file explorer with Textual.

https://github.com/darrenburns/kupo

**Reply by marcbelmont** [reply]
awesome, I'm checking it out.

**update:** it's not compatible with the textual main branch, and I couldn't figure out which one it's using. Editing a few lines here and there, I managed to run it, but it wasn't very functional. Anyway, it is similar to what I'm trying to build, and I'll look more into it.

**Reply by darrenburns** [reply]
There are two versions of it - I suspect you were running the old one. The newer stuff was in new_app.py which has way more stuff (opening files in editor, creating, copying, deleting (be careful 😄), multi selecting files). It's also unlikely to work on the latest version of Textual though.

**Reply by marcbelmont** [reply]
Yes, new_app + latest Textual has some issues. For example, `Message` seems to work differently now. But I quite like what you built so far!

**Comment by pbelange** [comment]
Hi! Jumping in as a long time user of Last File Manager (lfm). Same, very interested in a textual file explorer! It's a game changer, especially when working on remote machines from ssh. Any plans to maintain and make kupo into a main project of textual?

**Reply by darrenburns** [reply]
Hey - no there are no plans for that. It's using a very old version of Textual and I'd suspect I'd be more likely to rewrite it.

**Reply by pbelange** [reply]
Ok thanks. In any case, just letting you know there's interest! ;)

**Reply by huiyeruzhou** [reply]
Hi! I'm trying to read a log file and render it using textual, the app does similar things to what the textview does in editor. I'm using RichLog to implement it. What bothered me is that the app will stuck in syntax highlighting when I open a huge file (about 1000+ lines, a vesy sensible size to a detailed log), besides I have to flush the log area every 5 seconds to get the latest content. I have considered an increnmental policy but it may not work when the whole file is overwritten. I wonder if you have thought about these problem during write an editor, maybe you can share some idea and that will be very helpful ! Thank for any help in advance!

**Reply by davep** [reply]
@huiyeruzhou You may find https://github.com/Textualize/toolong of interest.

**Comment by huiyeruzhou** [comment]
> 嘿 - 不，没有这方面的计划。它使用的是非常旧的文本版本，我怀疑我更有可能重写它。



> @huiyeruzhou您可能会发现感兴趣的[https://github.com/Textualize/toolong 。](https://github.com/Textualize/toolong)

That's Cool! Thanks for recommending me such a great project!

**Comment by derVedro** [comment]
Not long ago, I stumbled upon [rovr](https://github.com/NSPC911/rovr) - a very fancy file explorer.

**Comment by kinthaiofficial** [comment]
Textual is actually a great fit for a file explorer — the built-in `DirectoryTree` widget gets you most of the way there, and the event system handles the interactive parts cleanly.

A minimal working file explorer:
```python
from textual.app import App, ComposeResult
from textual.widgets import DirectoryTree, Header, Footer
from textual.binding import Binding

class FileExplorer(App):
    BINDINGS = [
        Binding("q", "quit", "Quit"),
        Binding("enter", "open_file", "Open"),
    ]
    
    def compose(self) -> ComposeResult:
        yield Header()
        yield DirectoryTree(".")
        yield Footer()
    
    def on_directory_tree_file_selected(
        self, event: DirectoryTree.FileSelected
    ) -> None:
        self.notify(f"Selected: {event.path}")
        # open in editor, preview in panel, etc.
```

**Features worth adding**:

**Split pane with preview**:
```python
from textual.widgets import Static
from textual.containers import Horizontal

class FileExplorer(App):
    def compose(self) -> ComposeResult:
        with Horizontal():
            yield DirectoryTree(".", id="tree")
            yield Static("", id="preview")
    
    def on_directory_tree_file_selected(self, event):
        try:
            content = event.path.read_text()
            self.query_one("#preview").update(content[:2000])
        except Exception:
            self.query_one("#preview").update("[binary file]")
```

**Keyboard navigation for vim users**: bind `j`/`k` to scroll the tree, `l` to expand, `h` to collapse.

**Filtering**: add an `Input` widget that filters the tree by filename pattern using `DirectoryTree.filter_paths`.

Are you building this as a standalone tool or as a component within a larger TUI application?
